### PR TITLE
Added binary search into firebase-database/lib/ui/firebase_list.dart …

### PR DESCRIPTION
**Pull request by manishkumar-dsi**
Added binary search into _indexForKey function (firebase-database/lib/ui/firebase_list.dart) instead of normal search

### CODE REVIEW

File: packages/firebase_database/lib/ui/firebase_list.dart [hunk 1]
```diff
@@ -85,10 +85,18 @@ class FirebaseList extends ListBase<DataSnapshot>
...
+    int min = 0;
+    int max = _snapshots.length;
+    int mid;
+    while (min < max) {
+      mid = min + ((max - min) >> 1);
+      if (key == _snapshots[mid].key) {
+        return mid;
+      }
+      if (key.compareTo(_snapshots[mid].key) > 0) {
+        min = mid;
```
Comment (mravn-google) [inline packages/firebase_database/lib/ui/firebase_list.dart]: The invariant of the `while` loop is that the wanted key, if present, is in the half-open interval [min..max[

Then we need to update `min` to `mid + 1` in line 97.

I think we get an infinite loop, if `mid == min`. Which happens if `max == min + 1` and the key is not present.

Writing binary search correctly is tricky enough to warrant a unit test.

Comment (manishkumar-dsi) [reply packages/firebase_database/lib/ui/firebase_list.dart]: Thanks, @mravn-google. It requires more understanding of Flutter unit test case for me (newbie in Flutter). I will come back on this.

Comment (mravn-google) [reply packages/firebase_database/lib/ui/firebase_list.dart]: @manishkumar-dsi Let us know if you would prefer us taking it from here.

Comment (mravn-google) [reply packages/firebase_database/lib/ui/firebase_list.dart]: @manishkumar-dsi A rudimentary unit test is work in progress: https://github.com/flutter/plugins/pull/459

Comment (manishkumar-dsi) [reply packages/firebase_database/lib/ui/firebase_list.dart]: Sure @mravn-google. Please go ahead. Thanks

**Comment by mravn-google** [comment]
The comment about `firebase_list` being sorted is misleading, and hence binary search makes no sense. Sorry about this confusion.

**Pull request closed without merging**